### PR TITLE
Oip2

### PR DESCRIPTION
**Pull request by jerome-olympus**
OIP2 draft.

### CODE REVIEW

File: OIPs/oip-2.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+### 6. Fee model and MOT usage
+
+## Background
+
+Olympus Labs is a blockchain financial ecosystem centered around the Olympus Protocol for the decentralized creation of cryptocurrency-based financial products.
+
+We provide a web portal, SDKs, and APIs for financial product creators from both tech and non-tech backgrounds to easily create financial products for cryptocurrency investors.
+
+Along with the development of the platform, it turns out the original architecture built before no longer supports the further development of the system, thus we need to have a new architecture of the financial component based, which is highly extensible, reusable and makes developing of financial products much easier.
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: We can refer to OIP-1 here

File: OIPs/oip-2.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+## Overall design
+
+The overall idea is to have the entire system in different layers, each layer is consisted of its own parts. We will have 4 layers built into the system.
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: - is consisted of > consists of
- 4 > four

File: OIPs/oip-2.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+#### Layer 1: MOT
+
+​ MOT is in the very core of the entire ecosystem which will be involved into all the modules / interactions.
+
+#### Layer 2: Core Components
+
+The components are highly reusable ones in all financial products / applications. It starts from the ExchangeComponent, RiskControlComponent, WhitelistComponent to WithdrawComponent, etc. These can be used from different kinds of indices or funds, also the further products on our road map, like lending, options and futures.
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: - Don't need to mention the specific components yet, might be confusing for the user if it's included here. We can use generalized terms such as "exchange components".
- can be used from > can be used for
- also > as well as

File: OIPs/oip-2.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+#### Layer 3: Financial Protocols
+
+​ This layer is focusing on more specific protocol level, it takes different components from layer 2 according to its own needs, just like building a house taking different parts. If you need to charge fees for your fund, just connect to a Chargeable component, and the fee amount is already calculated for you. This makes it so easy to create financial protocols with only having to focus on the specific business logic instead of worrying about how to build the fundamental components like selling tokens, whitelist, etc.
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: - taking different parts > takes different parts
- Maybe the Chargeable is too specific here
- components like > components such as

File: OIPs/oip-2.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+![Developer](../assets/Developer.png)
+
+After the fund template becomes available on Olympus Template available, on the Fund manager portal, the fund managers will be able to see its information including descriptive name, supported features, etc. They can choose this template to create their own fund instance after certain confirmation depending on how the fund provides. The fund manager will use the portal to do the operations such as buying / selling / withdrawal, etc.
+
+![Fund Manager](../assets/Fund-Manager.png)
+
+When the fund instance is deployed to Ethereum, it registers to Olympus Marketplace, thus it's also available to its potential investors from different clients such as Olympus App / Portal or 3rd-party implementations like wallets. The investors can choose to invest on this fund instance by simply clicking the buy button on the client app / portal, then it interacts with the fund instance itself. Thanks for the base template, it implements the ERC20 standards, so investor automatically gets his fund token back to the address he specified.
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: Thanks for > Thanks to

File: OIPs/oip-2.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+![Investor](../assets/Investor.png)
+
+## Example workflow
+
+Let's take a look what happens from system's respective from different user cases to show how the Fund instance communicates with the core components.
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: perspective?
perspective for different use cases

File: OIPs/oip-2.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+## Fee model and MOT usage
+
+MOT plays a very important role in the new system, it's designated to be the currency in Olympus ecosystem. Without fee model, the system is not healthy to continue developing.
+
+Each component has its own fee model, base on practice, there are different model for different types of components.
+
+1.  Base on transaction amount.
+2.  Base on subscription.
+3.  Base on calls.
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: Based

File: OIPs/oip-2.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        public returns(bool success);
+
+    function sellTokens(string _exchangeId, ERC20[] _tokens, uint[] _amounts, uint[] _rates)
+        public returns(bool success);
+}
+```
+
+## Fee model and MOT usage
+
+MOT plays a very important role in the new system, it's designated to be the currency in Olympus ecosystem. Without fee model, the system is not healthy to continue developing.
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: Elaborate on why the system needs the fee model, what are the advantages in the form of incentives for developers/users/etc

File: OIPs/oip-2.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+​ This layer is focusing on more specific protocol level, it takes different components from layer 2 according to its own needs, just like building a house taking different parts. If you need to charge fees for your fund, just connect to a Chargeable component, and the fee amount is already calculated for you. This makes it so easy to create financial protocols with only having to focus on the specific business logic instead of worrying about how to build the fundamental components like selling tokens, whitelist, etc.
+
+#### Layer 4: Financial Applications
+
+​ Outside of the protocol is the applications layer. This is facing to organizations / end users, it includes Olympus App, Financial management portal, a DEX or a marketplace of financial products. It connects to the protocol to perform certain operations to serve customer's needs.
+
+The diagram is described as below:
+
+![Application Layers](../assets/Application-Layers.jpg)
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: Capitalize Portal, Lending, Marketplace and DEX

File: OIPs/oip-2.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+3.  Fund investors.
+
+Firstly, a financial smart contract developer sees the opportunity of developing financial products in Olympus ecosystem, so he uses his expertise of finance combining with the highly reusable components and the base templates provided by OlympusLabs to create his own Fund product, and then he can register his fund template to Olympus Template gallery. Developer can decide what he wants for his fee part if any fund manager wants to use this to create their financial products.
+
+![Developer](../assets/Developer.png)
+
+After the fund template becomes available on Olympus Template available, on the Fund manager portal, the fund managers will be able to see its information including descriptive name, supported features, etc. They can choose this template to create their own fund instance after certain confirmation depending on how the fund provides. The fund manager will use the portal to do the operations such as buying / selling / withdrawal, etc.
+
+![Fund Manager](../assets/Fund-Manager.png)
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: - Arrows back should not be a function, but like a variable. So getParameters() -> parameters
- Etherenum > Ethereum
- Marketplacce > Marketplace

File: OIPs/oip-2.md [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+When the fund instance is deployed to Ethereum, it registers to Olympus Marketplace, thus it's also available to its potential investors from different clients such as Olympus App / Portal or 3rd-party implementations like wallets. The investors can choose to invest on this fund instance by simply clicking the buy button on the client app / portal, then it interacts with the fund instance itself. Thanks for the base template, it implements the ERC20 standards, so investor automatically gets his fund token back to the address he specified.
+
+![Investor](../assets/Investor.png)
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: - Etherenum > Ethereum
- Returning arrows should not be function calls but parameters

File: OIPs/oip-2.md [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+## Example workflow
+
+Let's take a look what happens from system's respective from different user cases to show how the Fund instance communicates with the core components.
+
+When the fund template is created, it can hold different components and when the fund instance is initialized, the fund communicates with the specific components depending on what the action is. The components are highly reusable and can serve multiple contracts. In this case, the sample fund template uses whitelist, risk-control, withdrawal and exchange components.
+
+![Example investement](../assets/Example-investement.png)
```
Comment (Gerdinand) [inline OIPs/oip-2.md]: - Withdrawl > Withdrawal (2 places)
- Retrive > Retrieve
- Manager withdraws Exchange component connects to 3rd party exchanges, Investor withdraws, it does not
- Calculate base on > Calculate based on
- Return arrows should not be functions but variables

Comment (abelbordonado) [reply OIPs/oip-2.md]: image

1. We eventually check once the total amount, so this arrow shall in middle

image

2. Missing arrow back (the flow shall end where started)

File: OIPs/oip-2.md (hunk 9 above)
Comment (jerome-olympus) [reply OIPs/oip-2.md]: There will be a new chart. this is the draft yet.

**Pull request merged**